src/Io/PsrMiddlewareAdapter.php

```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace FrameworkX\Io;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface as PsrMiddlewareInterface;
+use React\Promise\PromiseInterface;
+use function React\Async\async;
+
+/**
+ * @internal
+ */
+class PsrMiddlewareAdapter
+{
+
+    /**
+     * @var PsrMiddlewareInterface
+     */
+    private $middleware;
+
+    public function __construct(PsrMiddlewareInterface $middleware) {
+        $this->middleware = $middleware;
+    }
+
+    /** @return PromiseInterface<ResponseInterface> */
+    public function __invoke(ServerRequestInterface $request, callable $next = null): PromiseInterface
+    {
+        return async(function () use ($request, $next) {
```

Comment:
async is not available for PHP < 7.4 but I wonder if that's really needed here.
Couldn't this be:
```php
$deferred = new Deferred();
$deferred->resolve($this->middleware->process($request, new PsrAwaitRequestHandler($next)));
return $deferred->promise();
```

or even just
```php
return $this->middleware->process($request, new PsrAwaitRequestHandler($next));
```
since PSR-15 middlewares will most probably be blocking anyways!?

Reply:
Don't think it's needed and like you said, with our commitment to PHP 7.4 support we can't really use it here for now (except building a workaround for PHP 7.4, but then we can just use the same workaround for all versions).